open_lm/model.py

```diff
@@ -125,7 +129,18 @@ def forward(self, x: torch.Tensor, is_causal=True):
 
         queries, keys, vals = self.pos_embed(queries, keys, vals)
 
-        output = self.attn_fn(queries, keys, vals, is_causal=is_causal)
+        if self.mup_base_n_heads:
+            '''
+            if mup_base_n_heads = 0, the parameter is not given and we do not scale.
+            If head_dim==mup_base_n_heads no scaling would be required.
+            mup requires to scale attention as 1/d instead of 1/sqrt(d), 
+            thus the relative scaling as below
+            '''
+            base_head_dim = self.mup_base_dim // self.mup_base_n_heads
+            multiplier = 1 / (self.head_dim / base_head_dim)
+            output = self.attn_fn(queries, keys, vals, is_causal=is_causal) * float(multiplier)
```

Comment:
I believe the scaling needs to be pre-softmax -- could be applied to the keys before `self.attn_fn`